File: src/main/java/tk/hiddenname/smarthome/repository/DeviceRepository.java
package tk.hiddenname.smarthome.repository;

import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.stereotype.Repository;
import tk.hiddenname.smarthome.entity.hardware.Device;
import tk.hiddenname.smarthome.entity.signal.SignalType;

import java.util.List;

@Repository
public interface DeviceRepository extends JpaRepository<Device, Integer> {

    List<Device> findAllByGpio_Type(SignalType type);

    List<Device> findAllByAreaId(Integer areaId);

    List<Device> findAllByGpio_TypeAndAreaId(SignalType type, Integer areaId);
}

File: src/main/java/tk/hiddenname/smarthome/repository/DeviceRepository.kt
package tk.hiddenname.smarthome.repository

import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.stereotype.Repository
import tk.hiddenname.smarthome.model.hardware.Device
import tk.hiddenname.smarthome.model.signal.SignalType
import javax.validation.constraints.NotNull

@Repository
interface DeviceRepository : JpaRepository<Device, Long> {

    fun findAllByGpioSignalType(gpioType: @NotNull SignalType): List<Device>

    fun findAllByAreaId(areaId: @NotNull Long): List<Device>

    fun findAllByGpioSignalTypeAndAreaId(type: @NotNull SignalType, areaId: @NotNull Long): List<Device>
}